### PR TITLE
Remove GDP var mover false positive warning

### LINKED ISSUE #496: Appropriate behavior for activate() and deactivate() for nested disjunctions?
```
from pyomo.environ import *
m = ConcreteModel()
m.d1 = Disjunct()
m.d2 = Disjunct()
m.d1.sub1 = Disjunct()
m.d1.sub2 = Disjunct()
m.d1.disj = Disjunction(expr=[m.d1.sub1, m.d1.sub2])
m.disj = Disjunction(expr=[m.d1, m.d2])
```

What should happen when disjuncts containing nested disjunctions are deactivated? `m.d1.deactivate()`.
The reclassifier hack complains about `m.d1.sub1` and `m.d1.sub2` not being expanded.

### PR DESCRIPTION
**Pull request by qtothec**
## Summary/Motivation:
Var mover complains about active disjuncts wherever they appear in the model, even if they are underneath a deactivated `Block` or deactivated `Disjunct` with `indicator_var` fixed to zero. This performs a check to make sure the disjunct in question is on an active path of the model before complaining.

Fixes #496 

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/gdp/plugins/gdp_var_mover.py [hunk 1]
```diff
@@ -91,3 +91,18 @@ def _apply_to(self, instance, **kwds):
...
+    def _disjunct_on_active_block(self, disjunct):
+        # Check first to make sure that the disjunct is not a
+        # descendent of an inactive Block or fixed and deactivated
+        # Disjunct, before raising a warning.
+        parent_block = disjunct.parent_block()
+        while parent_block is not None:
+            if parent_block.type() is Block and not parent_block.active:
+                return False
+            elif (parent_block.type() is Disjunct and not parent_block.active
+                  and parent_block.indicator_var.value == 0):
```
Comment (jsiirola) [inline pyomo/gdp/plugins/gdp_var_mover.py]: I think you also need to test that `not parent_block.indicator_var.active`.  The block could have been deactivated by a transformation, and the indicator_var could have a current (initial) value of 0, but the `Disjunct` is still technically "active".

Comment (qtothec) [reply pyomo/gdp/plugins/gdp_var_mover.py]: Ah, I see your point. There really needs to be a better way of indicating that a disjunct has been relaxed. I think that I should check:
```
if parent_block.type() is Block and not parent_block.active and not hasattr(parent_block, 'indicator_var'):
    return False
elif (parent_block.type() is Disjunct and not parent_block.active
       and parent_block.indicator_var.value == 0 and parent_block.indicator_var.fixed):
```
Does that seem right?

Comment (qtothec) [reply pyomo/gdp/plugins/gdp_var_mover.py]: Should we be using that in the reclassify transformation, then?

Comment (qtothec) [reply pyomo/gdp/plugins/gdp_var_mover.py]: Note to self: @jsiirola said yes to this question walking by.

**Comment by jsiirola** [comment]
This should be fine to merge pending successful test result.

**Comment by qtothec** [comment]
@jsiirola fyi, I changed the `gdp.reclassify` transformation to use `GDP_Error`, and now other tests are failing.

**Comment by qtothec** [comment]
@jsiirola Is it wrong to be using `GDP_Error` rather than `gdp.reclassify` in this case? If so, I need to figure out how to test for the existence of a `logger.error()` call.

**Comment by qtothec** [comment]
@jsiirola Ok, this is ready to look at again.

File: pyomo/gdp/tests/test_reclassify.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    def test_deactivated_parent_disjunct(self):
+        m = ConcreteModel()
+        m.d1 = Disjunct()
+        m.d1.sub1 = Disjunct()
+        m.d1.sub2 = Disjunct()
+        m.d1.disj = Disjunction(expr=[m.d1.sub1, m.d1.sub2])
+        m.d1.deactivate()
+        TransformationFactory('gdp.reclassify').apply_to(m)
+        # This should not raise an error
```
Comment (jsiirola) [inline pyomo/gdp/tests/test_reclassify.py]: I somewhat loathe "tests" that have no assertions.  Can you check that the appropriate disjunctions were all ended up with the correct ctypes?  I believe you want:
```
self.assertIs(m.d1.type(), Disjunct)
self.assertIs(m.d1.sub1.type(), Disjunct)
self.assertIs(m.d1.sub2.type(), Disjunct)
```

File: pyomo/gdp/tests/test_reclassify.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+    def test_deactivated_parent_block(self):
+        m = ConcreteModel()
+        m.d1 = Block()
+        m.d1.sub1 = Disjunct()
+        m.d1.sub2 = Disjunct()
+        m.d1.disj = Disjunction(expr=[m.d1.sub1, m.d1.sub2])
+        m.d1.deactivate()
+        TransformationFactory('gdp.reclassify').apply_to(m)
+        # This should not raise an error
```
Comment (jsiirola) [inline pyomo/gdp/tests/test_reclassify.py]: I somewhat loathe "tests" that have no assertions.  Can you check that the appropriate disjunctions were all ended up with the correct ctypes?  I believe you want:
```
self.assertIs(m.d1.type(), Block)
self.assertIs(m.d1.sub1.type(), Disjunct)
self.assertIs(m.d1.sub2.type(), Disjunct)
```

File: pyomo/gdp/tests/test_reclassify.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            TransformationFactory('gdp.reclassify').apply_to(m)
+
+    def test_active_parent_block(self):
+        m = ConcreteModel()
+        m.d1 = Block()
+        m.d1.sub1 = Disjunct()
+        m.d1.sub2 = Disjunct()
+        m.d1.disj = Disjunction(expr=[m.d1.sub1, m.d1.sub2])
+        with self.assertRaises(GDP_Error):
+            TransformationFactory('gdp.reclassify').apply_to(m)
```
Comment (jsiirola) [inline pyomo/gdp/tests/test_reclassify.py]: Can we add a few extra tests that show that for this model transforming the Disjunction (with the explicit `targets=` flag will relax the Disjuncts and that reclassify will correctly (and silently) convert them to Blocks?

**Comment by qtothec** [comment]
@jsiirola The behavior is not quite what you expect, since reclassify is really a transformation post-process that acts on the whole model.

**Pull request merged**